crates/ruma-identifiers/src/matrix_uri.rs

```diff
@@ -0,0 +1,467 @@
+                    let room_id = <&RoomOrAliasId>::try_from(second.as_ref())?;
+                    let event_id = <&EventId>::try_from(first.as_ref())?;
+                    Ok((room_id, event_id).into())
+                }
+                _ => Err(MatrixIdError::UnknownIdentifierPair.into()),
+            }
+        } else {
+            let id = percent_decode_str(s).decode_utf8()?;
+
+            match id.as_bytes()[0] {
+                b'@' => Ok(<&UserId>::try_from(id.as_ref())?.into()),
+                b'!' => Ok(<&RoomId>::try_from(id.as_ref())?.into()),
+                b'#' => Ok(<&RoomAliasId>::try_from(id.as_ref())?.into()),
+                b'$' => Err(MatrixIdError::MissingRoom.into()),
+                _ => Err(MatrixIdError::UnknownIdentifier.into()),
+            }
+        }
+    }
+
+    /// Construct a string with sigils from `self`.
+    ///
+    /// The identifiers will start with a sigil and be percent encoded.
+    ///
+    /// For events, the room ID or alias and the event ID will be separated by
+    /// a slash.
+    pub(crate) fn to_string_with_sigil(&self) -> String {
+        match self {
+            Self::Room(room_id) => percent_encode(room_id.as_bytes(), TO_ENCODE).to_string(),
+            Self::RoomAlias(room_alias) => {
+                percent_encode(room_alias.as_bytes(), TO_ENCODE).to_string()
+            }
+            Self::User(user_id) => percent_encode(user_id.as_bytes(), TO_ENCODE).to_string(),
+            Self::Event { room_id, event_id } => [
+                percent_encode(room_id.as_bytes(), TO_ENCODE).to_string(),
+                percent_encode(event_id.as_bytes(), TO_ENCODE).to_string(),
+            ]
+            .join("/"),
+        }
+    }
+}
+
+impl From<&RoomId> for MatrixId {
+    fn from(room_id: &RoomId) -> Self {
+        Self::Room(room_id.into())
+    }
+}
+
+impl From<&RoomAliasId> for MatrixId {
+    fn from(room_alias: &RoomAliasId) -> Self {
+        Self::RoomAlias(room_alias.into())
+    }
+}
+
+impl From<&UserId> for MatrixId {
+    fn from(user_id: &UserId) -> Self {
+        Self::User(user_id.into())
+    }
+}
+
+impl From<(&RoomOrAliasId, &EventId)> for MatrixId {
+    fn from(ids: (&RoomOrAliasId, &EventId)) -> Self {
+        Self::Event { room_id: ids.0.into(), event_id: ids.1.into() }
+    }
+}
+
+impl From<(&RoomId, &EventId)> for MatrixId {
+    fn from(ids: (&RoomId, &EventId)) -> Self {
+        Self::Event { room_id: <&RoomOrAliasId>::from(ids.0).into(), event_id: ids.1.into() }
+    }
+}
+
+impl From<(&RoomAliasId, &EventId)> for MatrixId {
+    fn from(ids: (&RoomAliasId, &EventId)) -> Self {
+        Self::Event { room_id: <&RoomOrAliasId>::from(ids.0).into(), event_id: ids.1.into() }
+    }
+}
+
+/// The `matrix.to` URI representation of a user, room or event.
+///
+/// Get the URI through its `Display` implementation (i.e. by interpolating it
+/// in a formatting macro or via `.to_string()`).
+#[derive(Debug, PartialEq, Eq)]
+pub struct MatrixToUri {
```

Comment:
Technically, this is a URL, and not a URI, i think it's easier to distinguish if its called `MatrixToUrl`